proof/infoflow/refine/RISCV64/Example_Valid_StateH.thy

```diff
@@ -3421,9 +3425,16 @@ lemma s0H_invs:
   apply (rule conjI)
    apply (clarsimp simp: irqs_masked'_def s0H_internal_def maxIRQ_def timer_irq_def irqInvalid_def)
   apply (rule conjI)
-   apply (clarsimp simp: valid_queues'_def obj_at'_def s0H_internal_def inQ_def)
-   apply (frule kh0H_dom_tcb)
-   apply (elim disjE, (clarsimp simp: kh0H_obj_def)+)[1]
+   apply (clarsimp simp: sym_heap_def opt_map_def projectKOs split: option.splits)
+   using kh0H_dom_tcb
+   apply (fastforce simp: kh0H_obj_def)
```

Comment:
Is it possible to avoid these `using` steps in the middle of the apply proof?

Reply:
I'll check. But is there anything wrong with using `using` in an apply proof? I thought that was absolutely fine. I try to avoid it, generally, but I have also been told that I should use `using` instead of `apply (insert blah)`

Reply:
One of the problems with `using` is that it screws with the autoindenter. The other issue is it's introducing a rule as an assumption, and then it's not clear what happens to it before it applies. Sometimes it can be avoided by `simp: whatever[simplified]`, sometimes with variable constraint/instantiation. I agree that `using` is better than `apply (insert blah)`.